### PR TITLE
B_reed/APPEALS-34071

### PR DESCRIPTION
**Pull request by breedbah**
Resolves [APPEALS-34071](https://jira.devops.va.gov/browse/APPEALS-34071)

# Description
-This PR introduces a new job, RefreshWebexAccessTokenJob, under the VirtualHearings module. This job is responsible for refreshing the access token used for Webex API calls.

Key changes include:

- The perform method creates a new instance of WebexService and calls its refresh_access_token method to get a new access token from the Webex API.

- If the response from the Webex API is successful, the new access and refresh tokens are stored in CredStash with the keys webex_#{Rails.deploy_env}_access_token and webex_#{Rails.deploy_env}_refresh_token respectively.

- If an error occurs during the process, it is caught and logged using the log_error method.

- This job is queued with low priority, indicating that it does not need to be run immediately and can wait until the system is less busy.

- This PR introduces a new job, RefreshWebexAccessTokenJob, under the VirtualHearings module. This job is responsible for refreshing the access token used for Webex API calls.

- Key changes include:

- The perform method creates a new instance of WebexService and calls its refresh_access_token method to get a new access token from the Webex API.

- If the response from the Webex API is successful, the new access and refresh tokens are stored in CredStash with the keys webex_#{Rails.deploy_env}_access_token and webex_#{Rails.deploy_env}_refresh_token respectively.

- If an error occurs during the process, it is caught and logged using the log_error method.

This job is queued with low priority, indicating that it does not need to be run immediately and can wait until the system is less busy.

## Acceptance Criteria
- [ ] Code compiles correctly

## Testing Plan

1. Go to [Jira Issue/Test Plan Link](https://jira.devops.va.gov/browse/JIRA-12345) or list them below

- [ ] For feature branches merging into master: Was this deployed to UAT?

## Integrations: Adding endpoints for external APIs
* [ ] Check that Caseflow's external API code for the endpoint matches the code in the relevant integration repo
  * [ ] Request: Service name, method name, input field names
  * [ ] Response: Check expected data structure
  * [ ] Check that calls are wrapped in MetricService record block
* [ ] Check that all configuration is coming from ENV variables
  * [ ] Listed all new ENV variables in description
  * [ ] Worked with or notified System Team that new ENV variables need to be set
* [ ] Update Fakes
* [ ] For feature branches: Was this tested in Caseflow UAT

# Best practices
## Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [ ] RSpec
- [ ] Jest
- [ ] Other

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [ ] No new code climate issues added

### Monitoring
- [ ] Are performance metrics (e.g., response time, throughput) being tracked?
- [ ] Are key application components monitored (e.g., database, cache, queues)?
- [ ] Is there a system in place for setting up alerts based on performance thresholds?

### Logging
- [ ] Are logs being produced at appropriate log levels (debug, info, warn, error, fatal)?
- [ ] Are logs structured (e.g., using log tags) for easier querying and analysis?
- [ ] Are sensitive data (e.g., passwords, tokens) redacted or omitted from logs?
- [ ] Is log retention and rotation configured correctly?
- [ ] Are logs being forwarded to a centralized logging system if needed?

### Auditing
- [ ] Are user actions being logged for audit purposes?
- [ ] Are changes to critical data being tracked ?
- [ ] Are logs being securely stored and protected from tampering or exposing protected data?

### Error Handling
- [ ] Are errors being caught and handled gracefully?
- [ ] Are appropriate error messages being displayed to users?
- [ ] Are critical errors being reported to an error tracking system (e.g., Sentry, ELK)?
- [ ] Are unhandled exceptions being caught at the application level ?

### Exception Handling
- [ ] Are custom exceptions defined and used where appropriate?
- [ ] Is exception handling consistent throughout the codebase?
- [ ] Are exceptions logged with relevant context and stack trace information?
- [ ] Are exceptions being grouped and categorized for easier analysis and resolution?

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -17,7 +17,7 @@ gem "bgs", git: "https://github.com/department-of-veterans-affairs/ruby-bgs.git"
 gem "bootsnap", require: false
 gem "browser"
 gem "business_time", "~> 0.9.3"
-gem "caseflow", git: "https://github.com/department-of-veterans-affairs/caseflow-commons", ref: "6377b46c2639248574673adc6a708d2568c6958c"
+gem "caseflow", git: "https://github.com/department-of-veterans-affairs/caseflow-commons", branch: "prototype/upgrade-aws-sdk"
```
Comment (ThorntonMatthew) [inline Gemfile:20]: We'll eventually need to get a PR open (that isn't for a `prototype/X` branch 😆) over in the caseflow-commons repo to push the AWS SDK upgrade into master prior to this going to production. It's fine for now.

File: app/jobs/virtual_hearings/refresh_webex_access_token_job.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+# This job is queued with low priority, indicating that it does not need to be run immediately
+# and can wait until the system is less busy.
+
+require "rcredstash"
+
+class VirtualHearings::RefreshWebexAccessTokenJob < CaseflowJob
+  queue_with_priority :low_priority
+
+  def perform
+    webex_service = WebexService.new(host: nil, port: nil, aud: nil, apikey: nil, domain: nil, api_endpoint: nil)
```
Comment (ThorntonMatthew) [inline app/jobs/virtual_hearings/refresh_webex_access_token_job.rb:23]:
```suggestion
    webex_service = WebexService.new(host: nil, port: nil, aud: nil, apikey: nil, domain: nil, api_endpoint: nil, query: nil)
```

Adding back in `query`.

File: app/services/external_api/webex_service.rb [hunk 3]
```diff
@@ -1,8 +1,29 @@
...
+# 4. The refresh_access_token method sends a POST request to the Webex API to refresh the access token.
+# 5. The fetch_recordings_list method sends a GET request to the Webex API to fetch a list of recordings.
+# 6. The fetch_recording_details method sends a GET request to the Webex API to fetch details of a recording.
+# 7. The send_webex_request method is a private method used to send requests to the Webex API
+#    with the specified body and method.
+#
+# All requests to the Webex API are recorded using the MetricsService.
+
 require "json"
+require "rcredstash"
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:21]: This can probably be removed since the `CredStash` class gets autoloaded at startup.

File: Gemfile (hunk 1 above)
Comment (breedbah) [reply Gemfile:20]: ok

File: app/jobs/virtual_hearings/refresh_webex_access_token_job.rb (hunk 2 above)
Comment (breedbah) [reply app/jobs/virtual_hearings/refresh_webex_access_token_job.rb:23]: good catch, i added it back to webex_service

File: app/services/external_api/webex_service.rb [hunk 4]
```diff
@@ -12,91 +33,105 @@ def initialize(host:, port:, aud:, apikey:, domain:, api_endpoint:, query:)
...
+    request.headers = headers
 
-    ExternalApi::WebexService::RecordingsListResponse.new(resp)
+    response = HTTPI.post(request)
+
+    ExternalApi::WebexService::AccessTokenRefreshResponse.new(response)
   end
 
-  def get_recording_details
+  def fetch_recordings_list
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:104]: Is this intentionally being changed in this branch?

File: app/services/external_api/webex_service.rb [hunk 5]
```diff
@@ -12,91 +33,105 @@ def initialize(host:, port:, aud:, apikey:, domain:, api_endpoint:, query:)
     @api_endpoint = api_endpoint
     @query = query
   end
+  # rubocop:enable Metrics/ParameterLists
 
-  def create_conference(conferenced_item)
+  def create_conference(virtual_hearing)
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:38]: Is this an intentional change?

If so, I think `hearing` may be better just because the links may be for a non-virtual hearing also.

File: app/services/external_api/webex_service.rb (hunk 4 above)
Comment (breedbah) [reply app/services/external_api/webex_service.rb:103]: Yes Marc suggested changing it. Its in another repo that Arianna was working on. Was causing github actions error

File: app/services/external_api/webex_service/response.rb [hunk 6]
```diff
@@ -41,6 +40,30 @@ def check_for_errors
   def error_message
     return "No error message from Webex" if resp.raw_body.empty?
 
-    JSON.parse(resp.raw_body).with_indifferent_access["message"]
+    begin
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service/response.rb:43]: Are these changes intentional?

Also, since this `ExternalApi::WebexService::Response` class is being used for both "thin" and "thick" client responses I do not know for sure that this logic will work for both APIs.

File: lib/fakes/webex_service.rb [hunk 7]
```diff
@@ -85,6 +85,10 @@ def get_recording_details
     )
   end
 
+  def refresh_access_token
```
Comment (ThorntonMatthew) [inline lib/fakes/webex_service.rb:88]: The return value differs from what the API will actually return. I've sent over a sample to you on VA Teams that shows the parsed response. I can get you one that's "raw" .

File: app/services/external_api/webex_service/response.rb (hunk 6 above)
Comment (breedbah) [reply app/services/external_api/webex_service/response.rb:43]: These were from the branch where Arianna made upates, I added them from her repo. I can change it back so that it doesnt cause issues

File: app/services/external_api/webex_service.rb [hunk 8]
```diff
@@ -12,91 +32,105 @@ def initialize(host:, port:, aud:, apikey:, domain:, api_endpoint:, query:)
...
-  def create_conference(conferenced_item)
+  def create_conference(hearing)
     body = {
       "jwt": {
-        "sub": conferenced_item.subject_for_conference,
-        "nbf": conferenced_item.nbf,
-        "exp": conferenced_item.exp
+        "sub": hearing.subject_for_conference,
+        "nbf": hearing.hearing.scheduled_for.beginning_of_day.to_i,
+        "exp": hearing.hearing.scheduled_for.end_of_day.to_i
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:42]: Also, `hearing.hearing`... This tells me that this method is assuming we're always going to pass in a `VirtualHearing` object, but I don't think we can make that assumption. This will need to be able to accept `Hearing` and `LegacyHearing` objects also.

Comment (ThorntonMatthew) [reply app/services/external_api/webex_service.rb:42]: Could both of those classes be made to implement these two methods?
https://github.com/department-of-veterans-affairs/caseflow/blob/b2eefa7c8d23eab69c6e1c9e021f99ea4ba26fd7/app/models/hearings/virtual_hearing.rb#L291-L297

Comment (ThorntonMatthew) [reply app/services/external_api/webex_service.rb:42]: They will need `subject_for_conference` too.

Comment (ThorntonMatthew) [reply app/services/external_api/webex_service.rb:42]: I don't think these are needed anymore since we don't create docket-level links anymore: https://github.com/department-of-veterans-affairs/caseflow/blob/b2eefa7c8d23eab69c6e1c9e021f99ea4ba26fd7/app/models/hearing_day.rb#L229-L239

File: app/services/external_api/webex_service.rb [hunk 9]
```diff
@@ -1,8 +1,28 @@
...
+# 6. The fetch_recording_details method sends a GET request to the Webex API to fetch details of a recording.
+# 7. The send_webex_request method is a private method used to send requests to the Webex API
+#    with the specified body and method.
+#
+# All requests to the Webex API are recorded using the MetricsService.
+
 require "json"
 
 class ExternalApi::WebexService
+  BASE_URL = "https://#{ENV['WEBEX_HOST_MAIN']}#{ENV['WEBEX_DOMAIN_MAIN']}"
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:23]: Removing this line in favor of placing the URL "pieces" definitions in the instantiation of the class, like here: https://github.com/department-of-veterans-affairs/caseflow/blob/aacde5b85eace7a4c54d4c639730ccaa0e1fb429/app/jobs/hearings/get_webex_recordings_details_job.rb#L47-L55

File: app/services/external_api/webex_service.rb [hunk 10]
```diff
@@ -12,91 +32,105 @@ def initialize(host:, port:, aud:, apikey:, domain:, api_endpoint:, query:)
     @api_endpoint = api_endpoint
     @query = query
   end
+  # rubocop:enable Metrics/ParameterLists
 
-  def create_conference(conferenced_item)
+  def create_conference(hearing)
     body = {
       "jwt": {
-        "sub": conferenced_item.subject_for_conference,
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:19]: Changing this back to what's on the feature/APPEALS-25121 branch.

File: app/services/external_api/webex_service.rb [hunk 11]
```diff
@@ -12,91 +32,105 @@ def initialize(host:, port:, aud:, apikey:, domain:, api_endpoint:, query:)
...
+    method = "GET"
+    resp = send_webex_request(body: body, method: method)
+    ExternalApi::WebexService::RecordingDetailsResponse.new(resp) if !resp.nil?
   end
 
   private
 
-  # :nocov:
-  def send_webex_request(body, method)
+  def send_webex_request(body: nil, method: nil)
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:119]: Will add a `header:` param at some point here.

File: app/services/external_api/webex_service.rb [hunk 12]
```diff
@@ -1,8 +1,28 @@
...
+#
+# All requests to the Webex API are recorded using the MetricsService.
+
 require "json"
 
 class ExternalApi::WebexService
+  BASE_URL = "https://#{ENV['WEBEX_HOST_MAIN']}#{ENV['WEBEX_DOMAIN_MAIN']}"
+
+  # rubocop:disable Metrics/ParameterLists
   def initialize(host:, port:, aud:, apikey:, domain:, api_endpoint:, query:)
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:26]: Are `host:` and `domain:` needed here? Either store it as a constant/env var depending if we find that the values change depending on environment.

File: app/services/external_api/webex_service.rb [hunk 13]
```diff
@@ -12,91 +32,105 @@ def initialize(host:, port:, aud:, apikey:, domain:, api_endpoint:, query:)
...
-
-    resp = send_webex_request(body, method)
+    resp = send_webex_request(body: body, method: "POST")
     return if resp.nil?
 
     ExternalApi::WebexService::CreateResponse.new(resp)
   end
 
-  def delete_conference(conferenced_item)
+  def delete_conference(hearing)
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:56]: Changing this back to what's on the feature/APPEALS-25121 branch.

File: app/services/external_api/webex_service.rb [hunk 14]
```diff
@@ -12,91 +32,105 @@ def initialize(host:, port:, aud:, apikey:, domain:, api_endpoint:, query:)
...
     body = {
       "jwt": {
-        "sub": conferenced_item.subject_for_conference,
+        "sub": hearing.subject_for_conference,
         "nbf": 0,
         "exp": 0
       },
       "aud": @aud,
+      "numGuest": 1,
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:64]: This could be removed.

File: app/services/external_api/webex_service.rb [hunk 15]
```diff
@@ -12,91 +32,105 @@ def initialize(host:, port:, aud:, apikey:, domain:, api_endpoint:, query:)
...
+    request.headers = headers
 
-    ExternalApi::WebexService::RecordingsListResponse.new(resp)
+    response = HTTPI.post(request)
+
+    ExternalApi::WebexService::AccessTokenRefreshResponse.new(response)
   end
 
-  def get_recording_details
+  def fetch_recordings_list
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service.rb:103]: Leave it as it stands in feature/APPEALS-25121

File: app/services/external_api/webex_service/response.rb [hunk 16]
```diff
@@ -41,6 +40,30 @@ def check_for_errors
...
+      body = JSON.parse(resp.raw_body)
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service/response.rb:44]: Leave it as it stands in feature/APPEALS-25121

File: lib/caseflow/error.rb [hunk 17]
```diff
@@ -424,7 +424,9 @@ class MeetingTypeNotFoundError < ConferenceCreationError; end
 
   class PexipApiError < ConferenceCreationError; end
   class PexipNotFoundError < PexipApiError; end
+  class WebexInvalidTokenError < ConferenceCreationError; end
```
Comment (ThorntonMatthew) [inline lib/caseflow/error.rb:427]: Can this one live with the other Webex errors?

File: lib/fakes/webex_service.rb [hunk 18]
```diff
@@ -53,7 +53,7 @@ def delete_conference(virtual_hearing)
     )
   end
 
-  def get_recordings_list
+  def fetch_recordings_list
```
Comment (ThorntonMatthew) [inline lib/fakes/webex_service.rb:56]: Should match what's on the feature branch.

File: lib/fakes/webex_service.rb [hunk 19]
```diff
@@ -69,7 +69,7 @@ def get_recordings_list
     )
   end
 
-  def get_recording_details
+  def fetch_recording_details
```
Comment (ThorntonMatthew) [inline lib/fakes/webex_service.rb:72]: Should match what's on the feature branch.

File: spec/jobs/virtual_hearings/refresh_webex_access_token_job_spec.rb [hunk 20]
```diff
@@ -0,0 +1,43 @@
+require "rails_helper"
+
+RSpec.describe VirtualHearings::RefreshWebexAccessTokenJob, type: :job do
+  let(:webex_service) { instance_double(Fakes::WebexService) }
```
Comment (ThorntonMatthew) [inline spec/jobs/virtual_hearings/refresh_webex_access_token_job_spec.rb:4]: This one can be removed too.

File: spec/services/external_api/webex_service_spec.rb [hunk 21]
```diff
@@ -43,6 +43,7 @@
             "exp": virtual_hearing.hearing.scheduled_for.end_of_day.to_i
           },
           "aud": aud,
+          "numGuest": 1,
```
Comment (ThorntonMatthew) [inline spec/services/external_api/webex_service_spec.rb:46]: This can probably be removed.

File: spec/services/external_api/webex_service_spec.rb [hunk 22]
```diff
@@ -96,6 +99,7 @@
             "exp": 0
           },
           "aud": aud,
+          "numGuest": 1,
```
Comment (ThorntonMatthew) [inline spec/services/external_api/webex_service_spec.rb:102]: This can probably be removed.

File: app/services/external_api/webex_service/response.rb [hunk 23]
```diff
@@ -41,6 +40,30 @@ def check_for_errors
...
+      if !invalid_token
+        {
+          message: body.dig(:message),
+          descriptions: body.dig(:errors)&.pluck(:description)&.compact
+        }
+      end
```
Comment (ThorntonMatthew) [inline app/services/external_api/webex_service/response.rb:50]: Please confirm that this code for error handling is what is in the feature branch.
